src/main/java/org/gradle/profiler/Main.java

```diff
@@ -61,21 +62,29 @@ public void run(String[] args) throws Exception {
             for (ScenarioDefinition scenario : scenarios) {
                 scenarioCount++;
                 Logging.startOperation("Running scenario " + scenario.getDisplayName() + " (scenario " + scenarioCount + "/" + totalScenarios + ")");
-
+                ScenarioInvoker<? extends ScenarioDefinition> invoker;
+                if (scenario instanceof BazelScenarioDefinition) {
+                    invoker = bazelScenarioInvoker;
+                } else if (scenario instanceof BuckScenarioDefinition) {
+                    invoker = buckScenarioInvoker;
+                } else if (scenario instanceof MavenScenarioDefinition) {
+                    invoker = mavenScenarioInvoker;
+                } else if (scenario instanceof GradleScenarioDefinition) {
+                    invoker = gradleScenarioInvoker;
+                } else {
+                    throw new IllegalArgumentException("Don't know how to run scenario.");
+                }
+                Consumer<BuildInvocationResult> resultConsumer = benchmarkResults.scenario(scenario, invoker.samplesFor(settings));
                 try {
-                    if (scenario instanceof BazelScenarioDefinition) {
-                        bazelScenarioInvoker.run((BazelScenarioDefinition) scenario, settings, benchmarkResults);
-                    } else if (scenario instanceof BuckScenarioDefinition) {
-                        buckScenarioInvoker.run((BuckScenarioDefinition) scenario, settings, benchmarkResults);
-                    } else if (scenario instanceof MavenScenarioDefinition) {
-                        mavenScenarioInvoker.run((MavenScenarioDefinition) scenario, settings, benchmarkResults);
-                    } else {
-                        gradleScenarioInvoker.run((GradleScenarioDefinition) scenario, settings, benchmarkResults);
-                    }
-
+                    invoker.run(scenario, settings, resultConsumer);
                 } catch (Throwable t) {
                     t.printStackTrace();
                     failures.add(t);
+                } finally {
+                    // write the results up to this point
```

Comment:
What does this mean? Does it write more results later?

Reply:
I've updated the comment. Hopefully it's clearer now.